---
title: "Analyze data using ArchR"
author: "Zhijian Li"
date: '`r format(Sys.Date(), "%Y-%B-%d")`'
output: html_document
---

```{r setup, include=FALSE}
library(ggplot2)
library(stringr)
library(magrittr)
library(WriteXLS)
library(tidyr)
library(dplyr)
library(plotly)
library(cluster)
library(cowplot)
library(gridExtra)
library(viridis)
library(GenomicRanges)
library(GenomeInfoDb)
library(data.table)
library(ArchR)
```

```{r set_parameters, echo=FALSE}
## set parameters
set.seed(42)
addArchRThreads(threads = 12)
addArchRGenome("hg38")
```

## Creating Arrow Files
```{r}
inputFiles <- c("CK100" = "../Alignment/CK100/outs/fragments.tsv.gz",
                "CK101" = "../Alignment/CK101/outs/fragments.tsv.gz",
                "CK102" = "../Alignment/CK102/outs/fragments.tsv.gz",
                "CK104" = "../Alignment/CK104/outs/fragments.tsv.gz",
                "CK105" = "../Alignment/CK105/outs/fragments.tsv.gz",
                "CK106" = "../Alignment/CK106/outs/fragments.tsv.gz",
                "CK107" = "../Alignment/CK107/outs/fragments.tsv.gz",
                "CK108" = "../Alignment/CK108/outs/fragments.tsv.gz",
                "CK109" = "../Alignment/CK109/outs/fragments.tsv.gz",
                "CK110" = "../Alignment/CK110/outs/fragments.tsv.gz",
                "CK111" = "../Alignment/CK111/outs/fragments.tsv.gz",
                "CK112" = "../Alignment/CK112/outs/fragments.tsv.gz",
                "CK130" = "../Alignment/CK130/outs/fragments.tsv.gz",
                "CK131" = "../Alignment/CK131/outs/fragments.tsv.gz",
                "CK132" = "../Alignment/CK132/outs/fragments.tsv.gz",
                "CK133" = "../Alignment/CK133/outs/fragments.tsv.gz")

filterTSS <- 4
filterFrags <- 3000

ArrowFiles <- createArrowFiles(
  inputFiles = inputFiles,
  sampleNames = names(inputFiles),
  outputNames = names(inputFiles),
  minTSS = filterTSS, 
  minFrags = filterFrags, 
  QCDir = "QualityControl",
  addTileMat = TRUE,
  addGeneScoreMat = TRUE
)

ArrowFiles
```

## Plotting
```{r, fig.width=6, fig.height=6}
for(sample in unique(names(inputFiles))){
    input_filename <- sprintf("./QualityControl/%s/%s-Pre-Filter-Metadata.rds", sample, sample)
    
    if(file.exists(input_filename)){
        Metadata <- readRDS(input_filename)
    
        ggtitle <- sprintf("%s\n%s\n%s",
            paste0(sample, "\nnCells Pass Filter = ", sum(Metadata$Keep)),
            paste0("Median Frags = ", median(Metadata$nFrags[Metadata$Keep==1])),
            paste0("Median TSS Enrichment = ", median(Metadata$TSSEnrichment[Metadata$Keep==1]))
          )
    
        gg <- ggPoint(
          x = pmin(log10(Metadata$nFrags), 5) + rnorm(length(Metadata$nFrags), sd = 0.00001),
          y = Metadata$TSSEnrichment + rnorm(length(Metadata$nFrags), sd = 0.00001), 
          colorDensity = TRUE,
          xlim = c(2.5, 5),
          ylim = c(0, max(Metadata$TSSEnrichment) * 1.05),
          baseSize = 6,
          continuousSet = "sambaNight",
          xlabel = "Log 10 (Unique Fragments)",
          ylabel = "TSS Enrichment",
          title = ggtitle,
          rastr = TRUE) + 
          geom_hline(yintercept=filterTSS, lty = "dashed", size = 0.25) +
          geom_vline(xintercept=log10(filterFrags), lty = "dashed", size = 0.25)
        
        print(gg)
    }
}
```


## Inferring Doublets
```{r, fig.width=6, fig.height=6}
doubScores <- addDoubletScores(
  input = ArrowFiles,
  k = 10, #Refers to how many cells near a "pseudo-doublet" to count.
  knnMethod = "UMAP", #Refers to the embedding to use for nearest neighbor search.
  LSIMethod = 1,
  outDir = "../data/DoubletScores"
)
```


## Creating an ArchRProject
```{r, fig.width=6, fig.height=6}
# With our Arrow files in hand, we are now ready to create an ArchRProject. An ArchRProject is associated with a set of Arrow files and is the backbone of nearly all ArchR analyses.
proj <- ArchRProject(
  ArrowFiles = ArrowFiles, 
  outputDirectory = "HF",
  showLogo = FALSE,
  copyArrows = TRUE #This is recommened so that you maintain an unaltered copy for later usage.
)

# Now we can filter putative doublets based on the previously determined doublet scores using the filterDoublets() function. This doesnâ€™t physically remove data from the Arrow files but rather tells the ArchRProject to ignore these cells for downstream analysis.
proj <- filterDoublets(ArchRProj = proj)


p1 <- plotGroups(ArchRProj = proj, 
                 groupBy = "Sample", 
                 colorBy = "cellColData", 
                 name = "TSSEnrichment",
                 alpha = 0.4,
                 plotAs = "violin",
                 addBoxPlot = TRUE)

p2 <- plotGroups(ArchRProj = proj, 
                 groupBy = "Sample", 
                 colorBy = "cellColData", 
                 name = "log10(nFrags)",
                 plotAs = "violin",
                 alpha = 0.4,
                 addBoxPlot = TRUE)

print(p1)
print(p2)
```

## Peak calling
```{r}
proj <- addGroupCoverages(ArchRProj = proj, 
                          groupBy = "Sample",
                          force = TRUE,
                          maxCells = 1000)

pathToMacs2 <- findMacs2()

proj <- addReproduciblePeakSet(
    ArchRProj = proj, 
    groupBy = "Sample", 
    pathToMacs2 = pathToMacs2
)

getPeakSet(proj)

proj <- addPeakMatrix(proj)
```

## add motif annotation
```{r}
proj <- addMotifAnnotations(ArchRProj = proj, 
                                 motifSet = "cisbp", 
                                 name = "Motif",
                                 force = TRUE)

proj <- addBgdPeaks(proj)

proj <- addDeviationsMatrix(
  ArchRProj = proj, 
  peakAnnotation = "Motif",
  force = TRUE
)

getAvailableMatrices(proj)
```


## save data
```{r}
saveArchRProject(ArchRProj = proj, 
                 load = FALSE)
```

## Session information
```{r}
sessionInfo()
```
